src/core/PWCharPool.cpp

```diff
@@ -542,3 +544,40 @@ bool CPasswordCharPool::CheckMasterPassword(const StringX &pwd, StringX &error)
     return false;
   }
 }
+
+double CPasswordCharPool::CalculatePasswordStrength(const StringX &password)
+{
+  if (password.empty()) {
+    return 0.0; // Empty password has no strength
+  }
+
+  // Determine the character pool size
+  std::unordered_set<char> charSet(password.begin(), password.end());
+  int poolSize = 0;
+
+  bool hasLower = false, hasUpper = false, hasDigits = false, hasSpecial = false;
+  for (char c : charSet) {
+    if (std::islower(c)) hasLower = true;
```

Comment:
A relatively easy improvement would be to keep separate sets for each class of characters and calculate the entropy based on the number of distinct characters in each set rather than the maximum set sizes.

Reply:
I've aligned the implementation with that of CPasswordCharPool::CheckMasterPassword, which uses additional criteria for password validation. CheckMasterPassword could also use the result of CalculatePasswordStrength. A comment in this function indicates the use of an entropy calculation.